src/state/controls/Button.ts

```diff
@@ -71,4 +71,11 @@ export class Button extends Control<IButtonData> implements IButton {
     public giveInput(input: IButtonInput): Promise<void> {
         return this.sendInput(input);
     }
+
+    public update(changedData: Partial<IButtonData>): Promise<void> {
+        if (changedData.cooldown) {
```

Comment:
>  Within control.update.

This is the issue, this is not possible. I mentioned above:

> Control&lt;T> has a method update(Partial&lt;T>) while Button extends Control&lt;R> while having a method update(Partial&lt;X>)

The [wikipedia page](https://en.wikipedia.org/wiki/Liskov_substitution_principle) on the LSP lists:

> Liskov's principle imposes some standard requirements on signatures... Contravariance of method arguments in the subtype.

Taking type `X` where `X` is not an extension or composition involving `R` breaks contravariance.